ethereumj-core/src/main/java/org/ethereum/sharding/processing/BeaconChainImpl.java

```diff
@@ -120,9 +124,14 @@ public synchronized ProcessingResult insert(Beacon block) {
         if ((vRes = beaconValidator.validateAndLog(block)) != ValidationResult.Success)
             return ProcessingResult.fromValidation(vRes);
 
-        // apply state transition
         Beacon parent = pullParent(block);
         BeaconState currentState = pullState(parent);
+        if ((vRes = proposerValidator.validateAndLog(block, parent, currentState)) != ValidationResult.Success)
+            return ProcessingResult.fromValidation(vRes);
+        if ((vRes = attestationsValidator.validateAndLog(block, parent, currentState)) != ValidationResult.Success)
```

Comment:
What if we merge `attestationsValidator` and `proposerValidator` into one validator with a couple of rules?